gemma-core/src/main/java/ubic/gemma/persistence/service/AbstractQueryFilteringVoEnabledDao.java

```diff
@@ -216,7 +215,30 @@ public long count( @Nullable Filters filters ) {
 
     @Override
     public long countWithCache( @Nullable Filters filters ) {
-        return doCountWithCache( filters, true );
+        return doCountWithCache( getByIdsFiltersIfPossible( filters ), true );
+    }
+
+    /**
+     * Optimize a filter by replacing it with a filter over IDs if possible.
+     * <p>
+     * This is taking advantage of the fact that we can cache the IDs of individual filters and intersect those results
+     * instead of running a complex database query.
+     *
+     * @param filters   the filters to be optimized
+     * @return an equivalent filter, optimized
+     */
+    private Filters getByIdsFiltersIfPossible( @Nullable Filters filters ) {
+        if ( filters != null ) {
+            Iterator<Iterable<Filter>> it = filters.iterator();
+            if ( it.hasNext() ) {
+                Set<Long> ids = new HashSet<>( doLoadIdsWithCache( Filters.by( IterableUtils.toList( it.next() ) ), null, true ) );
```

Comment:
It might not be worth to do this for every single filter.

Maybe having a list of properties that are expensive to filter by would be interesting.